src/lib.rs

```diff
@@ -293,6 +293,77 @@ pub fn resourceprep(s: &str) -> Result<Cow<'_, str>, Error> {
     Ok(Cow::Owned(normalized))
 }
 
+fn x520_mapped_to_nothing(c: char) -> bool {
+    if c.is_control() {
+        return true;
+    }
+    match c {
+        '\u{00AD}' | '\u{1806}' | '\u{034F}' | '\u{180B}'..='\u{180D}' |
+        '\u{FE00}'..='\u{FE0F}' | '\u{FFFC}' | '\u{200B}' => true,
+        _ => false,
+    }
+}
+
+/// Prepares a string according to the procedures described in Section 7 of
+/// [ITU-T Recommendation X.520 (2019)](https://www.itu.int/rec/T-REC-X.520-201910-I/en).
+///
+/// Note that this function does _not_ remove leading, trailing, or inner
+/// spaces as described in Section 7.6, because the characters needing removal
+/// will vary across the matching rules and ASN.1 syntaxes used.
+pub fn x520prep(s: &str, case_fold: bool) -> Result<Cow<'_, str>, Error> {
+    if s.chars().all(|c| matches!(c, ' '..='~') && (!case_fold || c.is_ascii_lowercase())) {
+        return Ok(Cow::Borrowed(s));
+    }
+
+    // 1. Transcode
+    // Already done because &str is enforced to be Unicode.
+
+    // 2. Map
+    let mapped = s.chars()
+        .filter(|&c| !x520_mapped_to_nothing(c))
+        .map(|c| if c.is_whitespace() { ' ' } else { c });
+
+    // 3. Normalize
+    let normalized = if case_fold {
```

Comment:
X.520 doesn't seem to describe the normalization step as optional.

Reply:
Correct. I misinterpreted what I thought was a mistake in the specification.

This sentence:

> For case ignore, numeric, and stored prefix string matching rules, characters are case-folded as per B.2 of IETF RFC 3454.

Made it sound conditional, but it just makes lowercasing conditional, which was my mistake here.